lightly/loss/ntx_ent_loss.py

```diff
@@ -136,37 +91,76 @@ def forward(self,
+        # negatives: shape: (embedding_size, memory_bank_size)
         out1, negatives = \
             super(NTXentLoss, self).forward(out1, update=out0.requires_grad)
 
+        # We use the cosine similarity, which is a dot product (einsum) here,
+        # as all vectors are already normalized to unit length.
+        # Notation in einsum: n = batch_size, c = embedding_size and k = memory_bank_size.
+
         if negatives is not None:
-            negatives = negatives.to(device)
             # use negatives from memory bank
-            l_pos = torch.einsum('nc,nc->n', [out0, out1]).unsqueeze(-1)
-            l_neg = torch.einsum('nc,ck->nk', [out0, negatives.clone().detach()])
+            negatives = negatives.to(device)
+
+
+            # sim_pos is of shape (batch_size, 1) and sim_pos[i] denotes the similarity
+            # of the i-th sample in the batch to its positive pair
+            sim_pos = torch.einsum('nc,nc->n', out0, out1).unsqueeze(-1)
+
+            # sim_neg is of shape (batch_size, memory_bank_size) and sim_neg[i,j] denotes the similarity
+            # of the i-th sample to the j-th negative sample
+            sim_neg = torch.einsum('nc,ck->nk', out0, negatives.clone().detach())
+
+
         else:
             # use other samples from batch as negatives
             output = torch.cat((out0, out1), axis=0)
-            similarity_matrix = self.similarity_function(output, output)
 
-            # filter out the scores from the positive samples
-            l_pos = torch.diag(similarity_matrix, batch_size)
-            r_pos = torch.diag(similarity_matrix, -batch_size)
-            l_pos = torch.cat([l_pos, r_pos]).view(2 * batch_size, 1)
+            # the similarity matrix has 4 quadrants:
+            # upper left: out0 to out0
+            # upper right: out0 to out1
+            # lower left: out1 to out0
+            # lower right: out1 to out1
+            similarity_matrix = torch.einsum("nc,mc->nm", output, output)
+
+            def concat_to_2x2(tensor00, tensor01, tensor10, tensor11):
+                tensor0 = torch.cat([tensor00, tensor01], dim=1)
+                tensor1 = torch.cat([tensor10, tensor11], dim=1)
+                tensor = torch.cat([tensor0, tensor1], dim=0)
+                return tensor
+
+            mask_base = torch.eye(batch_size, dtype=torch.bool, device=out0.device)
+            mask_0 = torch.zeros_like(mask_base, device=out0.device)
+
+            # the positive samples are on the diagonals of the upper right and lower left quadrant
+            mask_pos = concat_to_2x2(mask_0, mask_base, mask_base, mask_0)
+            sim_pos = similarity_matrix[mask_pos].view(2 * batch_size, -1)
+
+            # the negative samples are everywhere except on the diagonals of the 4 quadrants
+            mask_neg = ~concat_to_2x2(mask_base, mask_base, mask_base, mask_base)
+            sim_neg = similarity_matrix[mask_neg].view(2 * batch_size, -1)
```

Comment:
I think this part is unnecessarily slow and complicated (e.g. you use the `concat_to_2x2` twice here).

I'd like to propose a much shorter implementation which doesn't require masks:

```python
if negatives is not None:
    negatives = negatives.to(device)
    sim_pos = torch.einsum('nc,nc->n', out0, out1).unsqueeze(-1)
    sim_neg = torch.einsum('nc,ck->nk', out0, negatives.clone().detach())
    logits = torch.cat([sim_pos, sim_neg], dim=1)
    labels = torch.zeros(logits.shape[0]).long()
else:
    output = torch.cat([out0, out1], dim=0)
    logits = torch.einsum('nc,mc->nm', output, output)
    labels = torch.arange(batch_size)
    labels = torch.cat([labels + batch_size, labels])
```

This should result in the same output as your current implementation and is probably more efficient because we don't need the masked selects. What do you think? :)

(I left away the comments in my example but you should obviously keep them)

Reply:
The code for `if negatives is not None` is equal to what we have now.
The code for the else part does not seem correct to me for following reasons:
1. The similarities of samples to themselves, i.e. the values on the main diagonal are also part of the calculation. In the paper, however, they weren't. https://github.com/lightly-ai/lightly/blob/05bb9bd6c93c537763e8be8dc1f776db7ef4b793/tests/loss/test_NTXentLoss.py#L55
2. Consequently, the unittests testing against the other implementation fail.

The exact implementation I used was the following:
```python
output = torch.cat([out0, out1], dim=0)
logits = torch.einsum('nc,mc->nm', output, output)/self.temperature
labels = torch.arange(batch_size)
labels = torch.cat([labels + batch_size, labels]).long()
```

Reply:
If you want to speed up, the mask_pos and mask_neg could be saved and reused.

Reply:
Hm you are right but strangely, in an experiment I did locally, the result from my calculations was the same as the one for your calculations.

Reply:
Why don't we just follow the SimCLR implementation?

https://github.com/google-research/simclr/blob/3ad6700c1b139ee18e43f73546b7263a710de699/objective.py#L70